File: main/src/main/java/com/sedmelluq/lavaplayer/core/source/youtube/DefaultYoutubeTrackDetails.java
package com.sedmelluq.lavaplayer.core.source.youtube;

import com.sedmelluq.lavaplayer.core.http.HttpInterface;
import com.sedmelluq.lavaplayer.core.info.track.AudioTrackInfo;
import com.sedmelluq.lavaplayer.core.info.track.AudioTrackInfoTemplate;
import com.sedmelluq.lavaplayer.core.source.youtube.format.LegacyAdaptiveFormatsExtractor;
import com.sedmelluq.lavaplayer.core.source.youtube.format.LegacyDashMpdFormatsExtractor;
import com.sedmelluq.lavaplayer.core.source.youtube.format.LegacyStreamMapFormatsExtractor;
import com.sedmelluq.lavaplayer.core.source.youtube.format.StreamingDataFormatsExtractor;
import com.sedmelluq.lavaplayer.core.source.youtube.format.YoutubeTrackFormatExtractor;
import com.sedmelluq.lavaplayer.core.tools.JsonBrowser;
import com.sedmelluq.lavaplayer.core.tools.Units;
import com.sedmelluq.lavaplayer.core.tools.exception.FriendlyException;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static com.sedmelluq.lavaplayer.core.tools.Units.DURATION_MS_UNKNOWN;
import static com.sedmelluq.lavaplayer.core.tools.exception.FriendlyException.Severity.COMMON;
import static com.sedmelluq.lavaplayer.core.tools.exception.FriendlyException.Severity.SUSPICIOUS;

public class DefaultYoutubeTrackDetails implements YoutubeTrackDetails {
  private static final Logger log = LoggerFactory.getLogger(DefaultYoutubeTrackDetails.class);

  private static final YoutubeTrackFormatExtractor[] FORMAT_EXTRACTORS = new YoutubeTrackFormatExtractor[] {
      new LegacyAdaptiveFormatsExtractor(),
      new StreamingDataFormatsExtractor(),
      new LegacyDashMpdFormatsExtractor(),
      new LegacyStreamMapFormatsExtractor()
  };

  private final String videoId;
  private final YoutubeTrackJsonData data;
  private final AudioTrackInfoTemplate template;

  public DefaultYoutubeTrackDetails(String videoId, YoutubeTrackJsonData data, AudioTrackInfoTemplate template) {
    this.videoId = videoId;
    this.data = data;
    this.template = template;
  }

  @Override
  public AudioTrackInfo getTrackInfo() {
    return loadTrackInfo();
  }

  @Override
  public List<YoutubeTrackFormat> getFormats(HttpInterface httpInterface, YoutubeSignatureResolver signatureResolver) {
    try {
      return loadTrackFormats(httpInterface, signatureResolver);
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }

  @Override
  public String getPlayerScript() {
    return data.playerScriptUrl;
  }

  private List<YoutubeTrackFormat> loadTrackFormats(
      HttpInterface httpInterface,
      YoutubeSignatureResolver signatureResolver
  ) {
    for (YoutubeTrackFormatExtractor extractor : FORMAT_EXTRACTORS) {
      List<YoutubeTrackFormat> formats = extractor.extract(data, httpInterface, signatureResolver);

      if (!formats.isEmpty()) {
        return formats;
      }
    }

    log.warn(
        "Video {} with no detected format field, response {} polymer {}",
        videoId,
        data.playerResponse.format(),
        data.polymerArguments.format()
    );

    throw new FriendlyException("Unable to play this YouTube track.", SUSPICIOUS,
        new IllegalStateException("No track formats found."));
  }

  private AudioTrackInfo loadTrackInfo() {
    JsonBrowser playabilityStatus = data.playerResponse.get("playabilityStatus");

    if ("ERROR".equals(playabilityStatus.get("status").text())) {
      throw new FriendlyException(playabilityStatus.get("reason").text(), COMMON, null);
    }

    JsonBrowser videoDetails = data.playerResponse.get("videoDetails");

    if (videoDetails.isNull()) {
      return loadLegacyTrackInfo();
    }

    TemporalInfo temporalInfo = TemporalInfo.fromRawData(
        videoDetails.get("isLiveContent").asBoolean(false),
        videoDetails.get("lengthSeconds")
    );

    return buildTrackInfo(videoId, videoDetails, temporalInfo);
  }

  private AudioTrackInfo loadLegacyTrackInfo() {
    JsonBrowser args = data.polymerArguments;

    if ("fail".equals(args.get("status").text())) {
      throw new FriendlyException(args.get("reason").text(), COMMON, null);
    }

    TemporalInfo temporalInfo = TemporalInfo.fromRawData(
        "1".equals(args.get("live_playback").text()),
        args.get("length_seconds")
    );

    return buildTrackInfo(videoId, args, temporalInfo);
  }

  private AudioTrackInfo buildTrackInfo(String videoId, JsonBrowser holder, TemporalInfo temporalInfo) {
    return YoutubeTrackInfoFactory.create(
        template,
        videoId,
        holder.get("author").text(),
        holder.get("title").text(),
        temporalInfo.durationMillis,
        temporalInfo.isActiveStream
    );
  }

  private static class TemporalInfo {
    public final boolean isActiveStream;
    public final long durationMillis;

    private TemporalInfo(boolean isActiveStream, long durationMillis) {
      this.isActiveStream = isActiveStream;
      this.durationMillis = durationMillis;
    }

    public static TemporalInfo fromRawData(boolean wasLiveStream, JsonBrowser durationSecondsField) {
      long durationValue = durationSecondsField.asLong(0L);
      // VODs are not really live streams, even though that field in JSON claims they are. If it is actually live, then
      // duration is also missing or 0.
      boolean isActiveStream = wasLiveStream && durationValue == 0;

      return new TemporalInfo(
          isActiveStream,
          durationValue == 0 ? DURATION_MS_UNKNOWN : Units.secondsToMillis(durationValue)
      );
    }
  }
}

File: main/src/main/java/lavaplayer/source/youtube/DefaultYoutubeTrackDetails.kt
package lavaplayer.source.youtube

import lavaplayer.source.youtube.format.LegacyAdaptiveFormatsExtractor
import lavaplayer.source.youtube.format.LegacyDashMpdFormatsExtractor
import lavaplayer.source.youtube.format.LegacyStreamMapFormatsExtractor
import lavaplayer.source.youtube.format.StreamingDataFormatsExtractor
import lavaplayer.tools.ExceptionTools
import lavaplayer.tools.FriendlyException
import lavaplayer.tools.FriendlyException.Severity.COMMON
import lavaplayer.tools.FriendlyException.Severity.SUSPICIOUS
import lavaplayer.tools.ThumbnailTools
import lavaplayer.tools.Units
import lavaplayer.tools.Units.DURATION_MS_UNKNOWN
import lavaplayer.tools.io.HttpInterface
import lavaplayer.tools.json.JsonBrowser
import lavaplayer.track.AudioTrackInfo
import mu.KotlinLogging

class DefaultYoutubeTrackDetails(
    private val videoId: String,
    private val data: YoutubeTrackJsonData
) : YoutubeTrackDetails {

    companion object {
        private val log = KotlinLogging.logger { }
        private val FORMAT_EXTRACTORS = listOf(
            LegacyAdaptiveFormatsExtractor(),
            StreamingDataFormatsExtractor(),
            LegacyDashMpdFormatsExtractor(),
            LegacyStreamMapFormatsExtractor()
        )
    }

    override val trackInfo: AudioTrackInfo
        get() = loadTrackInfo()

    override val playerScript: String?
        get() = data.playerScriptUrl

    override fun getFormats(
        httpInterface: HttpInterface,
        signatureResolver: YoutubeSignatureResolver
    ): List<YoutubeTrackFormat> {
        try {
            return loadTrackFormats(httpInterface, signatureResolver)
        } catch (e: Exception) {
            throw ExceptionTools.toRuntimeException(e)
        }
    }

    private fun loadTrackFormats(
        httpInterface: HttpInterface,
        signatureResolver: YoutubeSignatureResolver
    ): List<YoutubeTrackFormat> {
        val formats = FORMAT_EXTRACTORS.firstNotNullOfOrNull {
            it.extract(data, httpInterface, signatureResolver)
                .takeUnless { formats -> formats.isEmpty() }
        }

        if (formats == null) {
            log.warn { "Video $videoId with no detected format field, response ${data.playerResponse.format()} polymer ${data.polymerArguments.format()}" }
            throw FriendlyException(
                "Unable to play this YouTube track.",
                SUSPICIOUS,
                IllegalStateException("No track formats found.")
            )
        }

        return formats
    }

    private fun loadTrackInfo(): AudioTrackInfo {
        val playabilityStatus = data.playerResponse["playabilityStatus"]
        if ("ERROR" == playabilityStatus["status"].safeText) {
            throw FriendlyException(playabilityStatus["reason"].text, COMMON, null)
        }

        val videoDetails = data.playerResponse["videoDetails"].takeUnless { it.isNull }
            ?: return loadLegacyTrackInfo()

        val temporalInfo =
            TemporalInfo.fromRawData(videoDetails["isLiveContent"].cast(false), videoDetails["lengthSeconds"])
        val artworkUrl = ThumbnailTools.extractYouTube(videoDetails, videoId)

        return buildTrackInfo(
            videoId,
            videoDetails["title"].safeText,
            videoDetails["author"].safeText,
            temporalInfo,
            artworkUrl
        )
    }

    private fun loadLegacyTrackInfo(): AudioTrackInfo {
        val args = data.polymerArguments
        if ("fail" == args["status"].safeText) {
            throw FriendlyException(args["reason"].safeText, COMMON, null)
        }

        val temporalInfo = TemporalInfo.fromRawData(args["live_playback"].safeText == "1", args["length_seconds"])
        val artworkUrl = ThumbnailTools.extractYouTube(args, videoId)

        return buildTrackInfo(videoId, args["title"].safeText, args["author"].safeText, temporalInfo, artworkUrl)
    }

    private fun buildTrackInfo(
        videoId: String,
        title: String,
        uploader: String,
        temporalInfo: TemporalInfo,
        artworkUrl: String
    ): AudioTrackInfo {
        return AudioTrackInfo(
            title,
            uploader,
            temporalInfo.durationMillis,
            videoId,
            "https://www.youtube.com/watch?v=$videoId",
            artworkUrl,
            temporalInfo.isActiveStream
        )
    }

    data class TemporalInfo(val isActiveStream: Boolean, val durationMillis: Long) {
        companion object {
            fun fromRawData(wasLiveStream: Boolean, durationSecondsField: JsonBrowser): TemporalInfo {
                val durationValue = durationSecondsField.cast(0L)
                // VODs are not really live streams, even though that field in JSON claims they are. If it is actually live, then
                // duration is also missing or 0.
                val isActiveStream = wasLiveStream && durationValue == 0L

                return TemporalInfo(
                    isActiveStream,
                    DURATION_MS_UNKNOWN.takeIf { durationValue == 0L } ?: Units.secondsToMillis(durationValue))
            }
        }
    }
}
